lib/clients/GcpClient.py

```diff
@@ -0,0 +1,553 @@
+            self.logger.error(message)
+            self.delete_snapshot(snapshot_name)
+            snapshot = None
+            raise Exception(message)
+
+        return snapshot
+
+    def _copy_snapshot(self, snapshot_id):
+        return self.get_snapshot(snapshot_id)
+
+    def _delete_snapshot(self, snapshot_id):
+        log_prefix = '[SNAPSHOT] [DELETE]'
+
+        try:
+            snapshot_deletion_operation = self.compute_client.snapshots().delete(
+                project=self.project_id, snapshot=snapshot_id).execute()
+
+            self._wait('Waiting for snapshot {} to be deleted...'.format(snapshot_id),
+                       (lambda operation_id, zonal_operation: self.get_operation_status(
+                           operation_id, zonal_operation) == 'DONE'),
+                       None,
+                       snapshot_deletion_operation['name'], False)
+
+            snapshot_exists = self.snapshot_exists(snapshot_id)
+
+            # Check if snapshot exists, if not then it is successfully deleted, else raise exception
+            if not snapshot_exists:
+                self._remove_snapshot(snapshot_id)
+                self.logger.info(
+                    '{} SUCCESS: snapshot-id={}'.format(
+                        log_prefix, snapshot_id))
+                return True
+            else:
+                message = '{} ERROR: snapshot-id={}, snapshot still exists'.format(
+                    log_prefix, snapshot_id)
+                self.logger.error(message)
+                raise Exception(message)
+        except Exception as error:
+            message = '{} ERROR: snapshot-id={}\n{}'.format(
+                log_prefix, snapshot_id, error)
+            self.logger.error(message)
+            raise Exception(message)
+
+    def _create_volume(self, size, snapshot_id=None):
+        log_prefix = '[VOLUME] [CREATE]'
+        volume = None
+
+        try:
+            disk_name = self.generate_name_by_prefix(self.disk_prefix)
+            disk_body = {
+                'name': disk_name,
+                'sizeGb': size,
+                'labels': self.tags
+                # default typ is pd-standard
+            }
+            if snapshot_id:
+                disk_body['sourceSnapshot'] = 'global/snapshots/{}'.format(
+                    snapshot_id)
+
+            self.logger.info('Creating disk with details {}'.format(disk_body))
+
+            disk_creation_operation = self.compute_client.disks().insert(
+                project=self.project_id, zone=self.availability_zone, body=disk_body).execute()
+
+            self._wait('Waiting for volume {} to get ready...'.format(disk_name),
+                       (lambda operation_id, zonal_operation: self.get_operation_status(
+                           operation_id, zonal_operation) == 'DONE'),
+                       None,
+                       disk_creation_operation['name'], True)
+
+            volume = self.get_volume(disk_name)
```

Comment:
If the volume/disk details are available from **disk_creation_operation** then avoid extra IaaS call made inside **self.get_volume(disk_name)**.

Reply:
disks.insert returns operation object which only provides information on the status of the operation. See [disk/insert](https://cloud.google.com/compute/docs/reference/rest/v1/disks/insert). Hence, we cannot avoid the extra call.